content/posts/using-the-redis-allocator-in-rust.md

```diff
@@ -0,0 +1,153 @@
+void REDISMODULE_API_FUNC(RedisModule_Free)(void *ptr);
+```
+
+These functions, like the rest of the Modules API, are defined as function pointers. When calling the functions from Rust, we need to dereference the function pointer first, which we do using the `unwrap()` method. We also need to do some casting to match up the pointer types. Finally, we need to use the `unsafe` keyword since we dereference raw pointers, which is not allowed in safe Rust for good reasons:
+
+```rust
+use std::alloc::{GlobalAlloc, Layout};
+use std::os::raw::c_void;
+
+struct RedisAlloc;
+
+unsafe impl GlobalAlloc for RedisAlloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        RedisModule_Alloc.unwrap()(layout.size()) as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        RedisModule_Free.unwrap()(ptr as *mut c_void)
+    }
+}
+```
+
+## The Crash
+
+Unfortunately, it's not that simple. When we build a module with this custom allocator and load it into Redis, it crashes on us. Redis does print a nice stack trace when it crashes, so let's look at it:
+
+```plaintext
+
+$ redis-server --loadmodule ./target/debug/examples/libhello.dylib
+thread panicked while processing panic. aborting.
+...
+Backtrace:
+0   redis-server                        0x000000010adce1dc logStackTrace + 110
+1   redis-server                        0x000000010adce562 sigsegvHandler + 236
+2   libsystem_platform.dylib            0x00007fff5b95db3d _sigtramp + 29
+3   ???                                 0x0000000000000000 0x0 + 0
+4   libhello.dylib                      0x000000010af5af6d _ZN3std9panicking18continue_panic_fmt17h0e74ab2b215a1401E + 157
+5   libhello.dylib                      0x000000010af5ae69 rust_begin_unwind + 9
+6   libhello.dylib                      0x000000010af6ea3f _ZN4core9panicking9panic_fmt17h09741a3213dba543E + 63
+7   libhello.dylib                      0x000000010af6e984 _ZN4core9panicking5panic17hb4bc64e7f35c9151E + 100
+8   libhello.dylib                      0x000000010af53108 _ZN4core6option15Option$LT$T$GT$6unwrap17h66957b4d942a4d3cE + 56
+9   libhello.dylib                      0x000000010af4f8f3 _ZN76_$LT$redis_module..alloc..RedisAlloc$u20$as$u20$core..alloc..GlobalAlloc$GT$5alloc17h6588ea2d7520a3ebE + 35
+...
+```
+
+So, it looks like we had a null pointer dereference here (`3 ??? 0x0000000000000000 0x0 + 0`), but what are all these weird symbols starting with `_ZN...`?
+
+A bit of searching turns up that this is the way Rust does name mangling: Unlike in C, and similarly to C++, in Rust multiple functions with the same name can coexist, since there are various namespace mechanisms such as modules and traits to distinguish them. To generate unique symbols that are C-compatible, the compiler mangles these to long and ugly unique names. To demangle these names back into the original, we can filter the output through [`rustfilt`](https://crates.io/crates/rustfilt). 
```

Comment:
Change to "After a bit of searching, we find that this is the way Rust does name mangling: ..."